src/blockchain/backend.rs

```diff
@@ -0,0 +1,145 @@
+use std::fs;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use anyhow::{anyhow, Result};
+use db_key::Key;
+use leveldb::database::Database;
+use leveldb::kv::KV;
+use leveldb::options::{Options, ReadOptions, WriteOptions};
+
+use crate::blockchain::header::{BackendHeader, HeaderMeta, HeaderMetaCache};
+use crate::blockchain::utils::{read_meta, wrap_with_level_db, wrap_with_rocks_db};
+use crate::blockchain::{utils, BackendDatabase, HeaderBackend};
+use crate::core::block::BlockId;
+use crate::core::header::Header;
+use crate::core::U256;
+
+pub struct BlockChainDB {
+    database: Arc<dyn BackendDatabase>,
+    meta: Arc<RwLock<ChainInfo>>,
+    header_meta_cache: Arc<HeaderMetaCache>,
+}
+
+impl BlockChainDB {
+    pub fn new(path: &Path) -> Result<BlockChainDB> {
+        let mut options = rocksdb::Options::default();
+        options.create_if_missing(true);
+        let database = rocksdb::DB::open(&options, path)?;
```

Comment:
Better to get BackendDatabase from outside, rather than initializing it here. This way we can choose between different backends. Also, I have a similar structure somewhere else. These two should be merged somehow